File: rn/Teacher/src/canvas-api/apis/pages.js
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import { paginate, exhaust } from '../utils/pagination'
import httpClient from '../httpClient'

export type PageParameters = {
  title: string,
  body: ?string,
  editing_roles: ?string,
  published: boolean,
  front_page: boolean,
}

export function getPages (courseID: string): Promise<ApiResponse<Page[]>> {
  const url = `courses/${courseID}/pages`
  const options = {
    params: {
      per_page: 99,
    },
  }
  let pages = paginate(url, options)
  return exhaust(pages)
}

export function getPage (courseID: string, url: string): Promise<ApiResponse<Page>> {
  return httpClient().get(`courses/${courseID}/pages/${url}`)
}

export function updatePage (courseID: string, url: string, parameters: PageParameters): Promise<ApiResponse<Page>> {
  return httpClient().put(`courses/${courseID}/pages/${url}`, { wiki_page: parameters })
}

export function createPage (courseID: string, parameters: PageParameters): Promise<ApiResponse<Page>> {
  return httpClient().post(`courses/${courseID}/pages`, { wiki_page: parameters })
}

export function deletePage (courseID: string, url: string): Promise<ApiResponse<Page>> {
  return httpClient().delete(`courses/${courseID}/pages/${url}`)
}

File: Core/Core/Pages/PageList/Pages.swift
//
// This file is part of Canvas.
// Copyright (C) 2020-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation

class Pages {
    let context: Context
    let env: AppEnvironment = .shared
    var handler: EmptyHandler?
    private var forceRefreshPages = false
    private var fetchingPages = false

    var frontPage: Store<GetFrontPage>?
    var all: Store<GetPages>?

    init(context: Context, handler: EmptyHandler?) {
        self.context = context
        self.handler = handler

        self.frontPage = env.subscribe(GetFrontPage(context: context)) { [weak self] in
            if self?.frontPage?.pending == false && self?.fetchingPages == false {
                self?.fetchingPages = true
                self?.frontPageDidUpdate()
            }
        }

        self.all = env.subscribe(GetPages(context: context)) { [weak self] in
            self?.pagesDidUpdate()
        }
    }

    func refresh(force: Bool = false) {
        //  refresh --> frontPage --> pages
        forceRefreshPages = force
        frontPage?.refresh(force: force)
    }

    func frontPageDidUpdate() {
        all?.refresh(force: forceRefreshPages)
    }

    func pagesDidUpdate() {
        let pending = all?.pending == true || frontPage?.pending == true
        if !pending {
            forceRefreshPages = false
            fetchingPages = false
            notify()
        }
    }

    func notify() {
        performUIUpdate {
            self.handler?()
        }
    }
}
